File: sudoku.py
#! /usr/bin/env python
import sys
import datetime
import glob
import re
import itertools

class Cell (object):
    def __init__(self, board, row, col):
        self.board = board
        self.row = row
        self.col = col
        self.possibles = '123456789'

    def __str__(self):
        return "[Cell:{0}:{1}:{2}]".format(self.row, self.col, self.possibles)

    def __repr__(self):
        # List uses __repr__ to print list contents
        return self.__str__()

    def box_number(self):
        return (self.row / 3) * 3 + (self.col / 3) 

    def has_possible(self, value):
        return len(self.possibles) > 1 and str(value) in self.possibles

    def is_subset_of(self, values):
        return len(set(self.possibles) - set(values)) == 0
                
    def remove_possibles(self, values, msg = None):
        found = False
        if len(self.possibles) > 1:
            for value in str(values):
                if value in self.possibles:
                    self.possibles = self.possibles.replace(value, '')
                    if msg:
                        print "{0}removed {1} from {2}:{3} => {4}".format(msg, value, self.row, self.col, self.possibles)
                    if len(self.possibles) == 1:
                        self.board.solve(self, self.possibles)
                    found = True
        return found
    
    
class Board (object):
    def __init__(self, data):
        self.cells = []
        for row in range(0,9):
            for col in range(0,9):
                self.cells.append(Cell(self, row, col))
        for row in range(0,9):
            for col in range(0,9):
                value = data[row * 9 + col]
                if re.match('[1-9]', value):
                    self.solve(self.get_cell(row, col), value)
        self.print_out()
        
                
    def print_out(self):
        data = []
        for c in self.cells:
            data.append(c.possibles)
        size = max(map(lambda x: len(x), data))

        data = map(lambda x: x.ljust(size), data)

        width = size * 9 + 9
        print '=' * width
        for row in range(0, 9):
            #print repr(data[row * 9: row * 9 + 9])
            print "|".join(data[row * 9: row * 9 + 9])
            if row in [2,5]:
                print '-' * width
        print '=' * width

    def get_cell(self, row, col):
        return self.cells[row * 9 + col]
        
    def get_row_text(self, row):
        row_data = []
        for cell in self.cells:
            if cell.row == row:
                if len(cell.possibles) == 1:
                    row_data.append(cell.possibles[0])
                else:
                    row_data.append('0')
        return "".join(row_data)
        
    def get_col_text(self, col):
        col_data = []
        for cell in self.cells:
            if cell.col == col:
                if len(cell.possibles) == 1:
                    col_data.append(cell.possibles[0])
                else:
                    col_data.append('0')
        return "".join(col_data)
        
    def get_box_text(self, row, col):
        data = []
        box = self.get_cell(row, col).box_number()
        for cell in self.cells:
            if cell.box_number() == box:
                if len(cell.possibles) == 1:
                    data.append(cell.possibles[0])
                else:
                    data.append('0')
        return "".join(data)
        
    def get_cells_by_row(self, row):
        cells = []
        for cell in self.cells:
            if cell.row == row:
                cells.append(cell)
        return cells

    def get_cells_by_col(self, col):
        cells = []
        for cell in self.cells:
            if cell.col == col:
                cells.append(cell)
        return cells

    def get_cells_by_box_number(self, box):
        cells = []
        for cell in self.cells:
            if cell.box_number() == box:
                cells.append(cell)
        return cells

    def solve(self, cell, value):
        """Solve a cell and remove any matching possibles on the same row, col, or box"""
        cell.possibles = value
        for c in self.cells:
            if c.row == cell.row or c.col == cell.col or c.box_number() == cell.box_number():
                c.remove_possibles(value) 
            
    def find_singles(self):
        """Find unique possibles by row, col, and box"""
        #print "========= Singles ============"
        for value in "123456789":
            for row in range(9):
                cells = filter(lambda x: x.has_possible(value), self.get_cells_by_row(row))
                if len(cells) == 1:
                    print "Single {0} in row {1}".format(value, row)
                    self.solve(cells[0], value)
                    return True
                            
            for col in range(9):
                cells = filter(lambda x: x.has_possible(value), self.get_cells_by_col(col))
                if len(cells) == 1:
                    print "Single {0} in col {1}".format(value, col)
                    self.solve(cells[0], value)
                    return True
                            
            for box in range(9):
                cells = filter(lambda x: x.has_possible(value), self.get_cells_by_box_number(box))
                if len(cells) == 1:
                    print "Single {0} in box {1}".format(value, box)
                    self.solve(cells[0], value)
                    return True
                            
        return False

    def possibles(self, cells):
        """ Return a list of all the unique outstanding possibles for a set of cells """
        possibles = map(lambda x: x.possibles if len(x.possibles) > 1 else "", cells)
        possibles = set("".join(possibles))
        return sorted(possibles)

    def combinations(self, possibles):
        """ all combinations of possibles from pairs to length - 1 """
        result = []
        for length in range(2,len(possibles)):
            result.extend(itertools.combinations(possibles, length))
        return result

    def naked_cells(self, cells):
        possibles = self.possibles(cells)
        combinations = self.combinations(possibles)
        for combo in combinations:
            matches = filter(lambda x: x.is_subset_of(combo), cells)
            if (len(combo) == len(matches)):
                found = False
                for cell in cells:
                    if not cell in matches:
                        if cell.remove_possibles(combo):
                            found = True
                if found:
                    print "Naked {0} in {1}".format(combo, matches)
                    return True

        return False

    def nakeds(self):
        for row in range(9):
            if self.naked_cells(self.get_cells_by_row(row)):
                return True
        for col in range(9):
            if self.naked_cells(self.get_cells_by_col(col)):
                return True
        for box in range(9):
            if self.naked_cells(self.get_cells_by_box_number(box)):
                return True
        return False

    def find_pointing_pairs(self):
        """If a possible occurs only twice in a box and these are on a row/col then the possible can be removed from the rest of the row/col"""
        #print "========= Pointing pairs ============"
        found = False
        for box in range(9):
            for value in "123456789":
                msg = "Pointing pair {0} in box {1} ".format(value, box)
                matches = filter(lambda x: x.has_possible(value), self.get_cells_by_box_number(box))
                cols = list(set(map(lambda x: x.col, matches)))
                rows = list(set(map(lambda x: x.row, matches)))
                if len(matches) in [2,3]:
                    if len(cols) == 1:
                        cells = set(self.get_cells_by_col(cols[0])) - set(matches)
                        for cell in cells:
                            if cell.remove_possibles(value, msg):
                                found = True
                        if found:
                            return True

                    if len(rows) == 1:
                        cells = set(self.get_cells_by_row(rows[0])) - set(matches)
                        for cell in cells:
                            if cell.remove_possibles(value, msg):
                                found = True
                        if found:
                            return True

        return False

    def box_line_reduction(self):
        """If a possible occurs in only one box of a row/col then it can be removed from the rest of the box.
        It is the reverse of pointing pairs
        """
        #print "========= Box Line Reduction ============"
        found = False
        for row in range(0, 9):
            for possible in range(0,9):
                boxes = []
                for cell in self.get_cells_by_row(row):
                    if cell.has_possible(possible):
                        box = cell.box_number()
                        if not box in boxes:
                            boxes.append(box)
                if len(boxes) == 1:
                    for cell in self.get_cells_by_box_number(boxes[0]):
                        if cell.row != row:
                            msg = "Box line reduction in box {0} row {1} for {2}: ".format(boxes[0], row, possible) 
                            if cell.remove_possibles(possible, msg):
                                found = True
                            
        for col in range(0, 9):
            for possible in range(0,9):
                boxes = []
                for cell in self.get_cells_by_col(row):
                    if cell.has_possible(possible):
                        box = cell.box_number()
                        if not box in boxes:
                            boxes.append(box)
                if len(boxes) == 1:
                    for cell in self.get_cells_by_box_number(boxes[0]):
                        if cell.col != col:
                            msg = "Box line reduction in box {0} col {1} for {2}: ".format(boxes[0], col, possible) 
                            if cell.remove_possibles(possible):
                                found = True
                            
        return found

    def x_wing(self):
        """If two rows have a possible in the same two columns then the possible can be removed from other rows on the same columns
        """
        #print "========= X-Wing ============"
        found = False
        for possible in range(1,10):
            rows = []
            for row in range(0, 9):
                cols = [possible, row]
                for cell in self.get_cells_by_row(row):
                    if cell.has_possible(possible):
                        cols.append(cell.col)
                if len(cols) == 4:
                    rows.append(cols)
    
            while len(rows) > 1:            
                for i in range(1, len(rows)):
                    first = str(rows[0][2]) + str(rows[0][3])
                    second = str(rows[i][2]) + str(rows[i][3])
                    if first == second:
                        col_one = rows[0][2]
                        col_two = rows[0][3]
                        row_one = rows[0][1]
                        row_two = rows[i][1]
                        msg = "X wing for {0} in rows {1} and {2}: ".format(possible, row_one, row_two) 
                        for cells in self.get_cells_by_col(col_one):
                            if cell.row != row_one and cell.row != row_two:
                                if cell.remove_possibles(possible, msg):
                                    found = True
                        for cells in self.get_cells_by_col(col_two):
                            if cell.row != row_one and cell.row != row_two:
                                if cell.remove_possibles(possible, msg):
                                    found = True
                        rows.pop(i)
                        break
                rows.pop(0)
                            
        for possible in range(1,10):
            cols = []
            for col in range(0, 9):
                rows = [possible, col]
                for cell in self.get_cells_by_col(col):
                    if cell.has_possible(possible):
                        rows.append(cell.row)
                if len(rows) == 4:
                    cols.append(rows)
    
            while len(cols) > 1:            
                for i in range(1, len(cols)):
                    first = str(cols[0][2]) + str(cols[0][3])
                    second = str(cols[i][2]) + str(cols[i][3])
                    if first == second:
                        row_one = cols[0][2]
                        row_two = cols[0][3]
                        col_one = cols[0][1]
                        col_two = cols[i][1]
                        msg = "X wing for {0} in cols {1} and {2}".format(possible, col_one, col_two) 
                        for cells in self.get_cells_by_row(row_one):
                            if cell.col != col_one and cell.col != col_two:
                                if cell.remove_possibles(possible, msg):
                                    found = True
                        for cells in self.get_cells_by_row(row_two):
                            if cell.col != col_one and cell.col != col_two:
                                if cell.remove_possibles(possible, msg):
                                    found = True
                        cols.pop(i)
                        break
                cols.pop(0)
                            
        return found

    def swordfish(self):
        """If a possible exists in the same 3 columns of 3 rows then the possible can be removed from other rows on the same columns
           Note: It doesn't need to be 3 - it could be 2.
        """
        #print "========= Swordfish ============"
        found = False
        for possible in range(1,10):
            rows = []
            for row in range(0, 9):
                cols = [possible, row]
                for cell in self.get_cells_by_row(row):
                    if cell.has_possible(possible):
                        cols.append(cell.col)
                if len(cols) == 4 or len(cols) == 5:
                    rows.append(cols)
            
            # for rows with 3 cols check if there are 2 other rows that are in those cols
            for row in range(0, len(rows)):
                if len(rows[row]) == 5:   # 3 cols
                    cols = rows[row][2:]
                    triple = [rows[row]]
                    for match in range(0, len(rows)):
                        if match != row:
                            is_match = True
                            for col in rows[match][2:]: 
                                if not col in cols:
                                    is_match = False
                            if is_match:
                                triple.append(rows[match])
                    if len(triple) == 3:
                        msg = "triple found " + repr(triple)
                        triple_rows = [triple[0][1], triple[1][1], triple[2][1]]
                        for col in cols:
                            for cell in self.get_cells_by_col(col):
                                if not cell.row in triple_rows:
                                    if cell.remove_possibles(possible, msg):
                                        found = True
                        
            
        return found

    def solved(self):
    
        for row in range(0, 9):
            answers = []
            for cell in self.get_cells_by_row(row):
                answers.append(cell.possibles)
            answers.sort()
            answer = "".join(answers)
            if answer != "123456789":
                return False
                
        for col in range(0, 9):
            answers = []
            for cell in self.get_cells_by_col(col):
                answers.append(cell.possibles)
            answers.sort()
            answer = "".join(answers)
            if answer != "123456789":
                return False
                
        for box in range(0, 9):
            answers = []
            for cell in self.get_cells_by_box_number(box):
                answers.append(cell.possibles)
            answers.sort()
            answer = "".join(answers)
            if answer != "123456789":
                return False
                
        return True

    def solution(self):
        solution = ""
        if self.solved:
            for row in range(0, 9):
                for col in range(0, 9):
                    solution += self.get_cell(row, col).possibles
            return solution
        else:
            return 'Not solved'
            
solutions = []
with open('top95expected.txt') as f:
    solutions = f.readlines()

problems = []
with open('top95.txt') as f:
    problems = f.readlines()

results = ""
index = 0
solved = 0
for problem in problems:
    index += 1
    if (len(sys.argv) == 2) and (sys.argv[1] != str(index)):
        continue

    board = Board(problem)

    running = True
    while running:
        if board.find_singles():
            pass
        elif board.nakeds():
            pass
        elif board.find_pointing_pairs():
            pass
        elif board.box_line_reduction():
            pass
        elif board.x_wing():
            pass
        elif board.swordfish():
            pass
        else:
            running = False

        board.print_out()
            
    if board.solved():
        print ">>> SOLVED >>>"
        expected = solutions[index-1].rstrip()
        solution = board.solution()
        if solution != expected:
            print "Houston we have a problem"
            print "Expected \n[{0}] \ngot\n[{1}]".format(expected, solution)
            sys.exit()
        solved += 1
        results += 'S'
    else:            
        print ">>> BEATS ME >>>"
        results += '.'
    if index % 10 == 0:
        results += '\n'

print results
print "Solved {0} of {1}".format(solved, index)
 

File: sudoku.rb
#!/usr/bin/env ruby

# Solves all 120 puzzles in Wayne Gould's Extreme Su Doku 2
# See http://www.sudokuwiki.org for explanation of strategies
# Note: I do not use hidden strategies since these seem to be mirror
# images of longer naked strategies and I personally am more likely
# to spot a long naked sequence than a short hidden one.

class Cell

    attr_reader :board, :row, :col, :possibles, :ref
    
    def initialize(board, row, col, ref)
        @board = board
        @row = row
        @col = col
        @ref = ref
        @possibles = '123456789'
    end
    
    def box()
        (@row / 3) * 3 + (@col / 3)
    end
    
    def has?(possible)
        @possibles.length > 1 and @possibles.include? possible
    end
    
    def solve(value)
        @possibles = value
    end
    
    def remove(value)
        found = false
        if @possibles.length > 1
            found = @possibles.sub!(value, '')
            if @possibles.length == 1
                @board.solve(self, @possibles)
            end
        end
        found
    end
    
    def remove_possibles(values)
        found = false        
        values.each_char { |value| found = true if remove(value) }
        found
    end
    
    def to_s()
        "Cell #{@ref} #{@row}:#{@col}=#{@possibles}"
    end
    
end 

class Board

    def initialize(data)
        @cells = []
        i = 0
        (0..8).each { |row| (0..8).each { |col| @cells << Cell.new(self, row, col, i); i += 1 }}
        i = 0
        (0..8).each { |row| (0..8).each { |col| 
            value = data[i,1]
            solve(cell(row, col), value) if '123456789'.include? value 
            i += 1
        }}
            
    end
    
    def to_s()
        rows = []
        (0..8).each do |row|
            rows << cells_by_row(row).each.map { |cell| cell.possibles }
        end
        (0..8).each do |col|
            max = 0
            (0..8).each do |row|
                max = rows[row][col].length if rows[row][col].length > max
            end
            (0..8).each do |row|
                rows[row][col] = rows[row][col] + " " * (max - rows[row][col].length)
            end
        end
        s = []
        s << '=' * 50
        i = 0
        rows.each do |row|
            i += 1
            s << "#{row.join("|")}"
            s << '-' * 50 if i == 3 or i == 6
        end
        s << '=' * 50
        s.join("\n")
    end
    
    def solve(solved_cell, value)
        solved_cell.solve(value)
        @cells.each { |cell| cell.remove(value) if cell.row == solved_cell.row or cell.col == solved_cell.col or cell.box == solved_cell.box }
    end
    
    def cell(row, col)
        @cells[row * 9 + col]
    end
    
    def cells_by_row(row)
        @cells.select { |cell| cell.row == row }
    end

    def cells_by_col(col)
        @cells.select { |cell| cell.col == col }
    end

    def cells_by_box(box)
        @cells.select { |cell| cell.box == box }
    end
    
    def cell_by_ref(ref)
        @cells.select { |cell| cell.ref == ref }[0]
    end
    
    def singles
        found = false
        
        (0..8).each do |row|
            "123456789".each_char do |possible|
                cells = cells_by_row(row).select { |cell| cell.has?(possible) }
                if cells.length == 1
                    puts "Single #{possible} in row #{row}"
                    solve(cells[0], possible)
                    found = true
                end
            end
        end
        
        (0..8).each do |col|
            "123456789".each_char do |possible|
                cells = cells_by_col(col).select { |cell| cell.has?(possible) }
                if cells.length == 1
                    puts "Single #{possible} in col #{col}"
                    solve(cells[0], possible)
                    found = true
                end
            end
        end
        
        (0..8).each do |box|
            "123456789".each_char do |possible|
                cells = cells_by_box(box).select { |cell| cell.has?(possible) }
                if cells.length == 1
                    puts "Single #{possible} in box #{box}"
                    solve(cells[0], possible)
                    found = true
                end
            end
        end
        
        found
    end

    def naked_cells( cells )
        found = false
        
        cells = cells.select { |cell| cell.possibles.length > 1 }
        possibles = cells.map { |cell| cell.possibles }.join.split(//).uniq.sort
        
        combos = []
        (2...possibles.length).each do |r|
            possibles.combination(r) { |x| combos << x }
        end
        
        combos.each do |combo|
            matches = cells.select { |cell| (cell.possibles.split(//) - combo).length == 0 }
            if matches.length == combo.length
                remove = combo.join
                cells.each do |cell|
                    if not matches.include? cell
                        found = true if cell.remove_possibles(remove)
                    end
                end
                if found
                    where = matches.map { |cell| "#{cell.row}:#{cell.col}" }.join(" ")
                    puts "Naked #{remove} found in #{where}"
                    return true
                end
            end
        end
        
        found
    end
    
    def nakeds
        # http://www.sudokuwiki.org/Naked_Candidates#NP
        (0..8).each { |row| return true if naked_cells(cells_by_row(row)) }
        (0..8).each { |col| return true if naked_cells(cells_by_col(col)) }
        (0..8).each { |box| return true if naked_cells(cells_by_box(box)) }
        false
    end
    
    def pointing_pairs
        # http://www.sudokuwiki.org/Intersection_Removal#IR
        found = false
        row = col = 0
        
        (0..8).each do |box|
            "123456789".each_char do |possible|
                pairs = cells_by_box(box).select { |cell| cell.has? possible}
                if pairs.length == 2 or pairs.length == 3
                    if pairs[0].row == pairs[1].row and pairs[0].row == pairs[-1].row
                        cells_by_row(pairs[0].row).each do |cell| 
                            if not pairs.include? cell and cell.remove(possible) 
                                found = true
                            end
                        end
                    end
                    if pairs[0].col == pairs[1].col and pairs[0].col == pairs[-1].col
                        cells_by_col(pairs[0].col).each do |cell| 
                            if not pairs.include? cell and cell.remove(possible) 
                                found = true
                            end
                        end
                    end
                    if found
                        puts "Pointing Pair #{possible} on box #{box}"
                        return true
                    end
                end
            end
        end
        
        found        
    end
        
    def box_line_reduction
        # http://www.sudokuwiki.org/Intersection_Removal#LBR
        found = false
        row = col = 0
        
        (0..8).each do |row|
            "123456789".each_char do |possible|
                cells = cells_by_row(row).select { |cell| cell.has? possible }
                boxes = cells.map { |cell| cell.box}.uniq
                if boxes.length == 1
                    cells_by_box(boxes[0]).each do |cell| 
                        if not cells.include? cell and cell.remove(possible) 
                            found = true
                        end
                    end
                end
                if found
                    puts "Box Line reduction of #{possible} on row #{row}"
                    return true
                end
            end
        end
        
        (0..8).each do |col|
            "123456789".each_char do |possible|
                cells = cells_by_col(col).select { |cell| cell.has? possible }
                boxes = cells.map { |cell| cell.box}.uniq
                if boxes.length == 1
                    cells_by_box(boxes[0]).each do |cell| 
                        if not cells.include? cell and cell.remove(possible) 
                            found = true
                        end
                    end
                end
                if found
                    puts "Box Line reduction of #{possible} on col #{col}"
                    return true
                end
            end
        end
        
        found        
    end
        
    def swordfishes( scale )
        # http://www.sudokuwiki.org/Sword_Fish_Strategy

        # Extension of swordfish to allow variations apart from 3x3
        # Note: 2x2 should be same as X-Wing

        type = "X-Wing" if scale == 2
        type = "Swordfish" if scale == 3
        type = "Swordfish-#{scale}" if scale > 3

        found = false
        matches = []

        "123456789".each_char do |possible|

            # Find rows with this possible. 
            # Store in a tuple [ possible, row, columns … ]

            rows = []        
            (0..8).each do |row|
                cols = [possible, row] + cells_by_row(row).select { |cell| cell.has? possible }.map { |cell| cell.col }
                rows << cols if cols.length >= 3 && cols.length <= (scale+2)
            end
            
            # If we have at least {scale} rows then try to find a combination of scale rows that share the same {scale} columns
            # First get all combinations of (scale) rows that contain our possible

            combos = []
            if rows.size >= scale
                rows.map {|row| row[1]}.combination(scale) { |x| combos << x }
            end
            
            # Now check if the (scale) rows share the same (scale) cols
            combos.each do |combo|
                scale_rows = rows.select {|row| combo.include? row[1] }
                cols = []
                scale_rows.each {|t| cols += t[2..-1]}
                cols.uniq!

                if scale_rows.size == scale && cols.size == scale
                    # Remove possibles from all the other rows
                    row_numbers = scale_rows.map {|t| t[1] }
                    cols.each do |col|
                        cells_by_col(col).each do |cell|
                            if not row_numbers.include? cell.row
                                found = true if cell.remove(possible)
                            end
                        end
                    end
                    if found
                        puts "#{type} #{possible} in rows #{row_numbers.inspect} cols #{cols.inspect}"
                        return true
                    end
                end
            end
        end
        
        "123456789".each_char do |possible|

            # Find cols with this possible. 
            # Store in a tuple [ possible, col, rows … ]


            cols = []        
            (0..8).each do |col|
                rows = [possible, col] + cells_by_col(col).select { |cell| cell.has? possible }.map { |cell| cell.row }
                cols << rows if rows.length >= 3 && rows.length <= (scale+2)
            end
            
            # If we have at least {scale} cols then try to find a combination of {scale} cols that share the same {scale} rows
            # First get all combinations of {scale} cols that contain our possible

            combos = []
            if cols.size >= scale
                cols.map {|col| col[1]}.combination(4) { |x| combos << x }
            end
            
            # Now check if the {scale} cols share the same {scale} rows
            combos.each do |combo|
                scale_cols = cols.select {|col| combo.include? col[1] }
                rows = []
                scale_cols.each {|t| rows += t[2..-1]}
                rows.uniq!
                if scale_cols.size == scale && rows.size == scale
                    # Remove possibles from all the other cols
                    col_numbers = scale_cols.map {|t| t[1] }
                    rows.each do |row|
                        cells_by_row(row).each do |cell|
                            if not col_numbers.include? cell.col
                                found = true if cell.remove(possible)
                            end
                        end
                    end
                    if found
                        puts "#{type} #{possible} in cols #{col_numbers.inspect} rows #{rows.inspect}"
                        return true
                    end
                end
            end
            
        end
        
        found        
    end
    
    def solved
        @cells.all? { |cell| cell.possibles.length == 1 }
    end

    def solution
        if solved
            return @cells.map { |cell| cell.possibles }.join
        else
            return "Not solved"
        end
    end
        
end

line = 0
solved=0
total=0
solutions = []
solution = ""


File.new("top95expected.txt", "r").each do |data|
   solutions << solution
end

solutions = File.readlines("top95expected.txt")

line = 0
File.new("top95.txt", "r").each do |data|
    exit if data.length < 81
    line += 1
    puts "=== Populating board #{line} ==="
    board = Board.new(data)
    running = true
    while running
        if board.singles
        elsif board.nakeds
        elsif board.pointing_pairs
        elsif board.box_line_reduction
        elsif board.swordfishes(2)
        elsif board.swordfishes(3)
        elsif board.swordfishes(4)
        else 
            running = false
        end
    end
    puts board.to_s
    if board.solved
        solved += 1
        expected = solutions[line-1].strip
        if board.solution != expected
            puts "Houston we have a problem … expected\n[#{expected}] \ngot\n[#{board.solution}]"
            exit
        end
        puts ">>> SOLVED >>>"
    else
        puts ">>> BEATEN >>>"
    end
    total += 1
end

puts ">>> Solved #{solved} of #{total} puzzles"



